gitpod-io/gitpod
# [ws-daemon] Refactor CPU limiting

## Description
This PR refactors the dynamic CPU limiting performed by ws-daemon. More details regarding motivation and solution can be found in this loom video: https://www.loom.com/share/0922bf11430c4ac084ebab6616e40b7e

This PR also adds a set of metrics to observe the new CPU limits:
- `gitpod_ws_daemon_cpulimit_workspaces_added_total`: Number of workspaces added to CPU control
- `gitpod_ws_daemon_cpulimit_workspaces_removed_total` Number of workspaces removed from CPU control
- `gitpod_ws_daemon_cpulimit_workspaces_burst_total`: Number of workspaces which received burst - `gitpod_ws_daemon_cpulimit_workspaces_cputime_seconds` CPU time of all observed workspaces
- `gitpod_ws_daemon_cpulimit_workspaces_throttled_total` Number of workspaces which ran with throttled CPU

## Related Issue(s)

Fixes #8035

## How to test
By default dynamic CPU limiting is disabled. It can be enabled using the experimental config in the installer, or by modifying the ws-daemon configmap. Therein, set the cpuLimiting.enabled to true, and configure the total available bandwidth on the node (e.g. 12 CPUs), limit and burst limit.

Then, start a few workspaces, observe their `/sys/fs/cgroup/cpu/cpu.cfs_quota_us` value while producing some CPU load.

In addition, one could try this version on a single VM and experiment with workspaces exhibiting different load profiles.

## Chaveats
- this PR removes the process priority management we had prior. It's unclear how much that actually helped, considering that it very outdated and focused on `bash` and `theia` and `gitpod-node` only. If we find user experience deteriorated too much by this removal, we'll want to bring it back in a future change.
- this change does not yet introduce QoS classes, but that would be straight forward to add down the road using annotations on the pod.
- this change removes the fixed CPU limiting that was in place as agent-smith action (limit CPU), and that was meant to afford paying customers more CPU. It's doubtful that the latter had much of an effect, but we'll want to follow up quickly with the QoS change.

## Release Notes

```release-note
Refactor dynamic CPU limiting to provide fairer scheduling.
```
